### PR TITLE
Don't omit default values when updating routing exclusions (#32721)

### PR DESCRIPTION
**Pull request by cbismuth**
Exclusion setting `cluster.routing.allocation.exclude._host` default value is an empty string.

* When an exclusion setting is sent with a null value the `o.e.c.s.Setting#innerGetRaw` API return an empty string (probably to avoid a `NullPointerException` to be raised).
* The `o.e.c.r.a.d.FilterAllocationDecider` class is developed to omit updates of default values for exclusion setting.
* That's why a null exclusion setting value is translated to an empty string which is equals to the exclusion default value which is configured to be ignored.

A simple fix would be to not omit default values for exclusion setting and keep the `NullPointerException` guard.

This is the purpose of this commit.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/settings/SettingTests.java
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/settings/SettingTests.java]: I don't think we should use real settings in SettingTests. You can create a dummy setting here that happens to be similar to a real setting.

**Review by rjernst — COMMENTED** [review]
I'll leave a review of the behavior changes for someone more familiar with these routing settings, but I have one comment on the test.

**Comment by cbismuth** [comment]
You're totally right @rjernst, thank you. I've updated my pull request.

Maybe @s1monw could help for this review, see [here](https://github.com/elastic/elasticsearch/commit/7b8d036ab5a8313843e637302d6a84bad3f5c7d7#diff-077710c48812c6301b3aea20709de601R105).

File: server/src/test/java/org/elasticsearch/common/settings/SettingTests.java
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/common/settings/SettingTests.java]: You're totally right @rjernst, thank you. I've updated my pull request.

**Comment by cbismuth** [comment]
I've referenced the wrong issue in the Git log message of my last commit, sorry for that noise.

File: server/src/test/java/org/elasticsearch/common/settings/SettingTests.java
Comment (jasontedor) [reply server/src/test/java/org/elasticsearch/common/settings/SettingTests.java]: I think that @rjernst suggestion was not about the value but about creating a dummy test setting (e.g., '`foo.bar`) that has similar properties to the routing settings. So rather than tying the test to a specific setting, we would want to be testing a generic setting with the properties of what you're trying to fix here.

Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/common/settings/SettingTests.java]: Got it! It would be far more better, yes. I'll working on it, thank you.

Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/common/settings/SettingTests.java]: I've updated my pull request and now both test cases are covered when setting a null value: defaults omitted or not.

**Comment by cbismuth** [comment]
Changes suggested by @rjernst and @jasontedor have been applied and test comments have been updated to ease GIVEN / WHEN / THEN comprehension.

PR is ready for review :+1:

**Comment by cbismuth** [comment]
Hi team, I have just updated this PR with `master` and rerun all tests, it's all green.

I have no doubt you're quite busy, I stay available for any necessary change.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: do we need these constants, they are only used in one place? Can we please just use the raw boolean?

**Review by s1monw — APPROVED** [review]
LGTM - I left one comment about the constants. Thanks for fixing this and sorry for the late reply. Can you fix this and then I pull it in.

**Comment by s1monw** [comment]
@cbismuth I looked at this again and I wonder if we should do the same for the other two as well? I mean wouldn't the same behavior be desired there too? And if that is the case should remove this boolean since this is the only user of it? WDYT?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java
Comment (cbismuth) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: Yes, I'll inline them as they aren't used in tests anymore, thank you.

**Comment by cbismuth** [comment]
Hi @s1monw, no worries. I think none of these three settings should omit default values, and as this boolean is not used anywhere else, as you suggested we should remove it.

I will add a commit to do so, so that we will be able to review such changes. I'll be able to revert it if you still prefer to keep this boolean parameter. I'll let you know when it's ready for review.

**Comment by cbismuth** [comment]
Hi @s1monw, PR is ready for review.

File: server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java
Comment (s1monw) [inline server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]:
```suggestion
```
this seems unnecessary

**Review by s1monw — APPROVED** [review]
LGTM left one comment. Thanks for the extra iteraton

**Comment by s1monw** [comment]
@elasticmachine ok to test

**Comment by s1monw** [comment]
@cbismuth our CI failed for this PR with this

```
FAILURE 0.01s J1 | ScopedSettingsTests.testAddConsumerAffixMap <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected null, but was:<[1]>
   >    at __randomizedtesting.SeedInfo.seed([23BCF16FE5CB6A8B:521C152BDDE41C44]:0)
   >    at org.elasticsearch.common.settings.ScopedSettingsTests.testAddConsumerAffixMap(ScopedSettingsTests.java:420)
   >    at java.lang.Thread.run(Thread.java:748)
```

`REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=23BCF16FE5CB6A8B -Dtests.class=org.elasticsearch.common.settings.ScopedSettingsTests -Dtests.method="testAddConsumerAffixMap" -Dtests.security.manager=true -Dtests.locale=et-EE -Dtests.timezone=Etc/GMT+9 -Dcompiler.java=11 -Druntime.java=8`

can you take a look

File: server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java
Comment (cbismuth) [reply server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: Good catch, thank you :+1:

**Comment by cbismuth** [comment]
Hi @s1monw, the build should be now fixed. Yesterday, tests passed on my local setup due to good/bad luck with random boolean testing.

As you suggested I've removed the unnecessary `omitDefaults` random boolean in tests (and I've also searched any other occurrence), it's all green on my side again.

**Comment by s1monw** [comment]
@cbismuth I will wait for CI to go green here on this PR and then merge it in! thank you!

**Comment by cbismuth** [comment]
Great! Thank you @s1monw :+1:

**Comment by cbismuth** [comment]
Build passed :+1:

**Comment by s1monw** [comment]
@cbismuth thanks for fixing this! I backported it to 6.x

**Comment by cbismuth** [comment]
Your welcome @s1monw, thank you for reviewing this PR!

**Pull request merged**
